### PR TITLE
Preliminary analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR is a first pass analysis of our batch experiment.

The question is: How does our ability to capture our real biological signal change as we vary the number of batch effects we have

There are 4 main notebooks:
0_process_data.ipynb:  Plot trends in the raw input data
1_train_vae.ipynb:  Train VAE using all gene expression data in the Pa compendium
2_simulate_data.ipynb:  Simulate gene expression data for 1 experiment using VAE latent space features
3_add_batch_effects.ipynb:  Add artificial batch effects to the simulated data
4_similarity_analysis.ipynb:  Calculate the similarity between the data representation of the original experimental data vs data with batch effects added.  

This work is hopefully going towards a PSB paper if results look promising.

### CODE REVIEW

File: scripts/nbconverted/2_simulate_data.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+# UMAP embedding of selected data
+input_data_UMAPencoded = umap.UMAP().fit_transform(data_selected)
+input_data_UMAPencoded_df = pd.DataFrame(data=input_data_UMAPencoded,
+                                         index=data_selected.index,
+                                         columns=['1','2'])
+
+
+g = ggplot(aes(x='1',y='2'), data=input_data_UMAPencoded_df) +             geom_point(alpha=0.5) +             scale_color_brewer(type='qual', palette='Set2') +             ggtitle("Input data")
+
+print(g)
```
Comment (ypar) [inline scripts/nbconverted/2_simulate_data.py]: Looks good!

File: scripts/nbconverted/3_add_batch_effects.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+seed(randomState)
+
+
+# In[2]:
+
+
+# Parameters
+analysis_name = 'treatment'
+num_simulations = 10
+num_batches = [1,2,3,4,5,6,7,8,9,10,15,20,25,30,35,40,45,50]
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: It may be useful to also think about nested batches for future directions. In realistic settings, 50 randomly spaced batches are less likely than a few big batches with subtle data clusters within each, for example.

Comment (ypar) [reply scripts/nbconverted/3_add_batch_effects.py]: Along the same line of thought, playing with different level of correlations between batches may be interesting.

File: scripts/nbconverted/3_add_batch_effects.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+
+# In[3]:
+
+
+# Create directories
+new_dir = os.path.join(os.path.dirname(os.getcwd()), "data", "batch_simulated")
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: consider using string formatting

File: scripts/nbconverted/3_add_batch_effects.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+else:
+    print('creating new directory: {}'.format(analysis_dir))
+os.makedirs(analysis_dir, exist_ok=True)
+
+
+# In[4]:
+
+
+# Load arguments
+simulated_data_file = os.path.join(
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: consider shortening this. see comments regarding paths in previous notebooks

File: scripts/nbconverted/3_add_batch_effects.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        index_col=0)
+    
+    # UMAP embedding of decoded batch data
+    batch_data_UMAPencoded = umap.UMAP().fit_transform(batch_data)
+    batch_data_UMAPencoded_df = pd.DataFrame(data=batch_data_UMAPencoded,
+                                             index=batch_data.index,
+                                             columns=['1','2'])
+    
+        
+    g = ggplot(aes(x='1',y='2'), data=batch_data_UMAPencoded_df) +                 geom_point(alpha=0.5) +                 scale_color_brewer(type='qual', palette='Set2') +                 ggtitle("{} Batches".format(i))
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: these look great! consider making the plot frame smaller so they don't take so much scrolling space, e.g.

`matplotlib.rcParams['figure.figsize'] = [width, height]`

or some individualized plot dim option

File: scripts/nbconverted/3_add_batch_effects.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+for i in num_batches:
+    batch_data_file = os.path.join(
+        os.path.dirname(os.getcwd()),
+        "data",
+        "batch_simulated",
+        analysis_name,
+        "Batch_"+str(i)+".txt")
+    
+    batch_data = pd.read_table(
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: good practice but really no need to multi-line it if the line is =<72 characters long

File: scripts/nbconverted/3_add_batch_effects.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    # Use trained model to encode expression data into SAME latent space
+    batch_data_PCAencoded = pca.fit_transform(batch_data)
+    
+    
+    batch_data_PCAencoded_df = pd.DataFrame(batch_data_PCAencoded,
+                                         index=batch_data.index,
+                                         columns=['1', '2'])
+    
+    g = ggplot(aes(x='1',y='2'), data=batch_data_PCAencoded_df)  +                 geom_point(alpha=0.5) +                 scale_color_brewer(type='qual', palette='Set2') +                 ggtitle("{} Batches".format(i))
```
Comment (ypar) [inline scripts/nbconverted/3_add_batch_effects.py]: looks good considering we are projecting only the two largest variance linear dimensions

File: scripts/nbconverted/4_similarity_analysis.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
+
+# coding: utf-8
+
+# # Similarity analysis
+# 
+# We want to determine if the different batch simulated data is able to capture the biological signal that is present in the original data:  How much of the real input data is captured in the simulated batch data?
```
Comment (ypar) [inline scripts/nbconverted/4_similarity_analysis.py]: curious. why svcca?

File: scripts/nbconverted/4_similarity_analysis.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    # SVCCA
+    svcca_results = cca_core.get_cca_similarity(original_data_UMAPencoded_df.T,
+                                          batch_data_UMAPencoded_df.T,
+                                          verbose=True)
+    
+    output_list.append(np.mean(svcca_results["cca_coef1"]))
+
+# Convert output to pandas dataframe
+svcca_umap_df = pd.DataFrame(output_list, columns=["svcca_mean_similarity"], index=num_batches)
+svcca_umap_df
```
Comment (ypar) [inline scripts/nbconverted/4_similarity_analysis.py]: may want to consider suppressing the stdout... it's taking up so much space with redundant messages...

File: scripts/nbconverted/4_similarity_analysis.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
```
Comment (ypar) [inline scripts/nbconverted/4_similarity_analysis.py]: they look almost identical with small, almost random variation. is it possible svcca is not the right tool for this?

Comment (ypar) [reply scripts/nbconverted/4_similarity_analysis.py]: or perhaps try simulation from raw data rather than normalized data?

File: scripts/nbconverted/4_similarity_analysis.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+    # SVCCA
+    svcca_results = cca_core.get_cca_similarity(original_data_PCAencoded_df.T,
+                                          batch_data_PCAencoded_df.T,
+                                          verbose=True)
+    
+    output_list.append(np.mean(svcca_results["cca_coef1"]))
+
+# Convert output to pandas dataframe
+svcca_pca_df = pd.DataFrame(output_list, columns=["svcca_mean_similarity"], index=num_batches)
+svcca_pca_df
```
Comment (ypar) [inline scripts/nbconverted/4_similarity_analysis.py]: so they are basically identical regardless of number of groups in comparison?

File: scripts/functions/helper_ae.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+# https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development
+import numpy as np
+import random as rn
+
+# The below is necessary in Python 3.2.3 onwards to
+# have reproducible behavior for certain hash-based operations.
+# See these references for further details:
+# https://docs.python.org/3.4/using/cmdline.html#envvar-PYTHONHASHSEED
+# https://github.com/keras-team/keras/issues/2280#issuecomment-306959926
+randomState = 123
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/functions/helper_ae.py]: This code was scraped from the blog that is referenced in order to try to get reproducible VAE models between training runs.  I am not exactly sure what each part of this is doing so that probably explains the redundancy.  If you think you have a better idea, I'm happy to sit down with you to do over it.

File: scripts/nbconverted/0_process_data.py [hunk 13]
```diff
@@ -0,0 +1,284 @@
...
+import random
+import glob
+import umap
+import matplotlib.pyplot as plt
+from ggplot import *
+
+from numpy.random import seed
+randomState = 123
+seed(randomState)
+
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: Thanks for the helpful tip!

File: scripts/nbconverted/0_process_data.py [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+# ## Data
+# 
+# Data downloaded from ADAGE repository [ADAGE](https://github.com/greenelab/adage).  
+# 
+# ```
+# data 
+# ```
+# 1. Raw data was downloaded from ArrayExpress using Affymetrix GeneChip
+# 2. Use [RMA](https://www.rdocumentation.org/packages/affy/versions/1.50.0/topics/rma) bioconductor library to convert raw array data to log 2 gene expression data.
+# 3. Only keep PA genes, remove control genes
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: This is a question I had myself.  The processing of the raw gene expression data originally came from Jie's paper and I tried to find more detail for what this meant, but I am still unclear.  I can try to ping Georgia and see if she knows though

File: scripts/nbconverted/0_process_data.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+from ggplot import *
+
+from numpy.random import seed
+randomState = 123
+seed(randomState)
+
+
+# ## Data
+# 
+# Data downloaded from ADAGE repository [ADAGE](https://github.com/greenelab/adage).
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: I will try to clarify this in my comment.  I downloaded the two datasets from the ADAGE repo, created by Jie: data and normalized_data.  I was just trying to describe what each of them were

File: scripts/nbconverted/0_process_data.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+normalized_data = pd.read_table(normalized_data_file, header=0, sep='\t', index_col=0).T
+normalized_data.head(5)
+
+
+# In[5]:
+
+
+# Read in metadata
+metadata = pd.read_table(metadata_file, header=0, sep='\t', index_col='ml_data_source')
+metadata
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: This was for debugging purposes, but I have shortened it now

File: scripts/nbconverted/0_process_data.py [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+    metadata_vector = metadata[metadata_field]
+
+    sample_id = metadata_selected.index
+
+    metadata_selected_labeled = metadata_selected.assign(
+        strain=(
+            list( 
+                map(
+                    lambda x:
+                    values[0] if x.lower().strip() == values[0].lower()
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: So its slightly different depending on the metadata field that is selected.  But if you have an idea for how to clean this section up, I'm happy to talk to you about it.

File: scripts/nbconverted/0_process_data.py [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+# Read in data
+normalized_data = pd.read_table(normalized_data_file, header=0, sep='\t', index_col=0).T
+normalized_data.head(5)
+
+
+# In[5]:
+
+
+# Read in metadata
+metadata = pd.read_table(metadata_file, header=0, sep='\t', index_col='ml_data_source')
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: This column is used, because it is the one that appears to contain the sample id that corresponds to the gene expression data file.  But you're correct that I am loosing some samples since there are values missing.  I can also double check with Georgia about this.

File: scripts/nbconverted/0_process_data.py [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+embedding_df['metadata'] = list(data_labeled[metadata_field])
+print(embedding_df.shape)
+embedding_df.head(5)
+
+
+# In[13]:
+
+
+# Plot
+ggplot(aes(x='1',y='2', color='metadata'), data=embedding_df) +         geom_point(alpha=0.5) +         scale_color_brewer(type='qual', palette='Set2')
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: The lines are broken up in the nb, not sure why its not showing up like that in the py file

File: scripts/nbconverted/0_process_data.py [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+embedding_df = pd.DataFrame(data=embedding, columns=['1','2'])
+embedding_df['metadata'] = list(data_labeled[metadata_field])
+print(embedding_df.shape)
+embedding_df.head(5)
+
+
+# In[13]:
+
+
+# Plot
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/0_process_data.py]: Do you have a palette you'd recommend?  This was the best one that distinguished between different classes

File: scripts/nbconverted/1_train_vae.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+        print('directory already exists: {}'.format(each_dir))
+    else:
+        print('creating new directory: {}'.format(each_dir))
+    os.makedirs(analysis_dir, exist_ok=True)
+
+
+# In[3]:
+
+
+# Parameters
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/1_train_vae.py]: These are parameters from previous times we've trained the VAE that seemed to give us a good loss curve

[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/1_train_vae.py]: Something I will be playing around with as a next step as discussed

File: scripts/nbconverted/1_train_vae.py [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+latent_dim = 300
+epsilon_std = 1.0
+num_PCs = latent_dim
+
+
+# In[4]:
+
+
+# Train nonlinear (VAE)
+vae.tybalt_2layer_model(learning_rate,
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/1_train_vae.py]: The stats are saved per epoch, this output is more to inform me of the progress in training and when it might be complete

File: scripts/nbconverted/2_simulate_data.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+randomState = 123
+seed(randomState)
+
+
+# In[2]:
+
+
+# Parameters
+analysis_name = 'treatment'
+experiment_id = 'E-GEOD-24036'
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/2_simulate_data.py]: Changed, as discussed

File: scripts/nbconverted/2_simulate_data.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+num_simulated_samples = 100
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply scripts/nbconverted/2_simulate_data.py]: Very good point.  Since its a variable I can change this value in the future.

File: scripts/nbconverted/3_add_batch_effects.py (hunk 2 above)
Comment (ajlee21) [reply scripts/nbconverted/3_add_batch_effects.py]: Good points.  Casey and I have discussed this as a future direction once we are able to get realistic simulated data generated.

File: scripts/nbconverted/3_add_batch_effects.py (hunk 5 above)
Comment (ajlee21) [reply scripts/nbconverted/3_add_batch_effects.py]: Eventually I had planned to stitch these individual plots into a single figure, but that's also for future work

File: scripts/nbconverted/3_add_batch_effects.py (hunk 6 above)
Comment (ajlee21) [reply scripts/nbconverted/3_add_batch_effects.py]: I realized that its a lot easier for me to read in this format as I've started using it and this way its also consistent

File: scripts/nbconverted/4_similarity_analysis.py (hunk 8 above)
Comment (ajlee21) [reply scripts/nbconverted/4_similarity_analysis.py]: This was a suggestion from Casey to allow us to compare the representation of the input data without batches vs some number of batches.  I definitely need to look into other possible options.

File: scripts/nbconverted/4_similarity_analysis.py (hunk 10 above)
Comment (ajlee21) [reply scripts/nbconverted/4_similarity_analysis.py]: I may be misunderstanding your idea here, but we're using normalized data because the VAE needs to have each feature (gene) to be in the same range to train.

**Comment by ajlee21** [comment]
I have update the files in this PR based on the above comments.  @ypar please let me know if there are additional comments or questions you have.

File: scripts/nbconverted/0_process_data.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ypar) [reply scripts/nbconverted/0_process_data.py]: you could cite jie's paper here too if that's the case.

File: scripts/nbconverted/1_train_vae.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (ypar) [reply scripts/nbconverted/1_train_vae.py]: OK that's fine. You might want to consider printing a progress bar instead to save scroll space and reduce redundant texts.

File: scripts/nbconverted/4_similarity_analysis.py (hunk 10 above)
Comment (ypar) [reply scripts/nbconverted/4_similarity_analysis.py]: as we discussed, this may be a byproduct of the generator not working properly so nullifying my comment here.

File: scripts/nbconverted/0_process_data.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ypar) [reply scripts/nbconverted/0_process_data.py]: because you are lowering alpha, generally i'd recommend non-pastel colors. e.g. set1 instead of set2. or something of that nature
